store/sqlstore/channel_store.go

```diff
@@ -3950,46 +3948,175 @@ func (s SqlChannelStore) UpdateSidebarCategories(userId, teamId string, categori
+		"UserId":       preference.UserId,
+		"ChannelId":    preference.Name,
+		"CategoryType": model.SidebarCategoryFavorites,
+	}
+	var query string
+	if s.DriverName() == model.DATABASE_DRIVER_MYSQL {
+		query = `
+			DELETE
+				SidebarChannels
+			FROM
+				SidebarChannels
+			JOIN
+				SidebarCategories ON SidebarChannels.CategoryId = SidebarCategories.Id
+			WHERE
+				UserId = :UserId
+				AND ChannelId = :ChannelId
+				AND Type = :CategoryType`
+	} else {
+		query = `
+			DELETE FROM
+				SidebarChannels
+			USING
+				SidebarCategories
+			WHERE
+				SidebarChannels.CategoryId = SidebarCategories.Id
+				AND SidebarChannels.UserId = :UserId
+				AND SidebarChannels.ChannelId = :ChannelId
+				AND SidebarCategories.Type = :CategoryType`
+	}
+
+	if _, err := transaction.Exec(query, params); err != nil {
+		return errors.Wrap(err, "Failed to remove sidebar entries for preference")
+	}
+
+	return nil
+}
+
+func (s SqlChannelStore) addChannelToFavoritesCategory(transaction *gorp.Transaction, preference *model.Preference) error {
+	if preference.Category != model.PREFERENCE_CATEGORY_FAVORITE_CHANNEL {
+		return nil
+	}
+
+	var channel *model.Channel
+	if obj, err := transaction.Get(&model.Channel{}, preference.Name); err != nil {
+		return errors.Wrapf(err, "Failed to get favorited channel with id=%s", preference.Name)
+	} else {
+		channel = obj.(*model.Channel)
+	}
+
+	var teamIdFilter sq.Sqlizer
+	if channel.TeamId == "" {
+		// Accept the Favorites category on any team
+		teamIdFilter = sq.Expr("(1=1)")
+	} else {
+		teamIdFilter = sq.Eq{"TeamId": channel.TeamId}
+	}
+
+	// Get the IDs of the Favorites category/categories that the channel needs to be added to
+	idsQuery, idsParams, _ := s.getQueryBuilder().
+		Select("Id").
+		From("SidebarCategories").
+		Where(sq.And{
+			teamIdFilter,
+			sq.Eq{
+				"UserId": preference.UserId,
+				"Type":   model.SidebarCategoryFavorites,
+			},
+			sq.Expr("Id NOT IN (select CategoryId from SidebarChannels where ChannelId = ?)", channel.Id),
+		}).ToSql()
+
+	var categoryIds []string
+	if _, err := transaction.Select(&categoryIds, idsQuery, idsParams...); err != nil {
+		return errors.Wrap(err, "Failed to get Favorites sidebar categories")
+	}
+
+	if len(categoryIds) == 0 {
+		// The channel is already in the Favorites category/categories
+		return nil
+	}
+
+	// For each category ID, insert a row into SidebarChannels with the given channel ID and a SortOrder that's less than
+	// all existing SortOrders in the category so that the newly favorited channel comes first
+	query, params, _ := s.getQueryBuilder().
+		Insert("SidebarChannels").
+		Columns(
+			"ChannelId",
+			"CategoryId",
+			"UserId",
+			"SortOrder",
+		).
+		Select(
+			sq.Select().
+				Column(sq.Expr("? as ChannelId", preference.Name)).
+				Column("SidebarCategories.Id as CategoryId").
+				Column(sq.Expr("? as UserId", preference.UserId)).
+				Column(sq.Expr("COALESCE(MIN(SidebarChannels.SortOrder) - 10, 0) as SortOrder")).
```

Comment:
Is `sq.Expr` required here?

Reply:
I'm not sure if it's actually necessary on any of these. Squirrel definitely has the ability to add a lot of redundant code if you're not careful...